### PR TITLE
add `CCLocal_storage` to containers-thread

### PR DESCRIPTION
**Pull request by c-cube**
Basic local storage emulation using an atomic-protected map. The goal is for this to work both with 4.xx threads, 5.xx domains, and 5.xx threads running on a given domain.

cc @bronsa :)

### CODE REVIEW

**Comment by c-cube** [comment]
A big design question is whether to pick:

- never empty:
  ```ocaml
  type 'a t

  val create : init:(unit -> 'a) -> unit -> 'a t

  val get : 'a t -> 'a
  ```

- can be empty (the current):
  ```ocaml
  type 'a t

  val create : unit -> 'a t

  val get : 'a t -> 'a option
  ```

**Comment by Bronsa** [comment]
> A big design question is whether to pick:
> 
>     * never empty:
>       ```ocaml
>       type 'a t
>       
>       val create : init:(unit -> 'a) -> unit -> 'a t
>       
>       val get : 'a t -> 'a
>       ```
> 
>     * can be empty (the current):
>       ```ocaml
>       type 'a t
>       
>       val create : unit -> 'a t
>       
>       val get : 'a t -> 'a option
>       ```

Why not both? 
```ocaml
type 'a t

val create : init:(unit -> 'a) -> unit -> 'a t

val create_empty : unit -> 'a option t

val get : 'a t -> 'a

```

**Comment by bluddy** [comment]
If deleting the thread's local storage is not part of the API, I think it makes sense to have it always contain a value.

On the other hand, what would the behavior be when you create a new thread and pass the handle to it? Can you now `get` on the TLS and expect a value? The init function can only be called when you `get`. Should it be called also on a `set` from the new thread?

File: src/threads/CCLocal_storage.mli [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+
+    Be aware that some threads might
+    have exited without cleaning up behind them. See {!with_} for
+    scope-protected modification of the variable that will cleanup
+    properly (like {!Fun.protect}).
+*)
+
+val with_ : 'a t -> 'a -> (unit -> 'b) -> 'b
+(** [with_ var x f] sets [var] to [x] for this thread, calls [f()], and
+    then restores the old value of [var] for this thread. *)
```
Comment (bluddy) [inline src/threads/CCLocal_storage.mli:43]: This seems like an odd `with_` function to me. Is it really useful?

**Comment by c-cube** [comment]
@bluddy if it's always full, which makes sense, then: `get v` would initialize `v` if it was not already initialized for this thread; `set v` would just directly write the new value. You'd never observe an absence of values.

**Comment by bluddy** [comment]
@c-cube I'm just concerned about possible mutable state issues. In a regular array, `init` happens at once and is a lot more straightforward. In our case here, it happens only when you observe it, and doesn't happen at all if you don't observe it.

**Comment by c-cube** [comment]
Yeah I mean, if you do side-effectful stuff in a TLS initializer and then complain, it's kind of your fault :D. Unlike `Array.init`, the number of instances here is not fixed, and new threads might be started later, so initialization _must_ be lazy.

**Comment by c-cube** [comment]
There's `thread-local-storage` already, and maybe it'll also land in the stdlib.

**Pull request closed without merging**